db.go

```diff
@@ -127,48 +163,95 @@ func (x *dummyAuthenticator) ResetPasswordFinish(identity string, token string,
 	}
 }
 
-func (x *dummyAuthenticator) CreateIdentity(identity, password string) error {
+func (x *dummyUserStore) CreateIdentity(identity, password string) error {
 	x.usersLock.Lock()
 	defer x.usersLock.Unlock()
-	if _, exists := x.users[CanonicalizeIdentity(identity)]; !exists {
-		x.users[CanonicalizeIdentity(identity)] = &dummyUser{identity, password, ""}
+	user := getUser(x.users, identity)
+	if user == nil {
+		userId := generateUserId(x.users)
+		x.users[userId] = &dummyUser{userId, identity, password, ""}
 		return nil
 	} else {
 		return ErrIdentityExists
 	}
 }
 
-func (x *dummyAuthenticator) RenameIdentity(oldIdent, newIdent string) error {
+func (x *dummyUserStore) RenameIdentity(oldIdent, newIdent string) error {
 	x.usersLock.Lock()
 	defer x.usersLock.Unlock()
 
-	if _, exists := x.users[CanonicalizeIdentity(newIdent)]; exists {
-		return ErrIdentityExists
-	}
-
-	oldIdent = CanonicalizeIdentity(oldIdent)
 	newKey := CanonicalizeIdentity(newIdent)
+	newUser := getUser(x.users, newKey)
+	if newUser == nil {
+		oldIdent = CanonicalizeIdentity(oldIdent)
 
-	if user, exists := x.users[oldIdent]; exists {
-		delete(x.users, oldIdent)
-		x.users[newKey] = &dummyUser{newIdent, user.password, ""}
-		return nil
+		oldUser := getUser(x.users, oldIdent)
+
+		if oldUser != nil {
+			delete(x.users, oldUser.userId)
+			x.users[oldUser.userId] = &dummyUser{oldUser.userId, newIdent, oldUser.password, ""}
+			return nil
+		} else {
+			return ErrIdentityAuthNotFound
+		}
 	} else {
-		return ErrIdentityAuthNotFound
+		return ErrIdentityExists
 	}
 }
 
-func (x *dummyAuthenticator) GetIdentities() ([]string, error) {
+func (x *dummyUserStore) GetIdentities() ([]string, error) {
 	x.usersLock.RLock()
 	list := []string{}
-	for identity, _ := range x.users {
-		list = append(list, identity)
+	for _, v := range x.users {
+		list = append(list, v.name)
 	}
 	x.usersLock.RUnlock()
 	return list, nil
 }
 
-func (x *dummyAuthenticator) Close() {
+func (x *dummyUserStore) GetIdentityFromUserId(userId UserId) (identity string, err error) {
+	x.usersLock.Lock()
+	defer x.usersLock.Unlock()
+
+	for _, v := range x.users {
+		if v.userId == userId {
+			identity = v.name
+			return
+		}
+	}
+	err = ErrIdentityAuthNotFound
+	return
+}
+
+func (x *dummyUserStore) GetUserIdFromIdentity(identity string) (userId UserId, err error) {
+	x.usersLock.Lock()
+	defer x.usersLock.Unlock()
+
+	for _, v := range x.users {
+		if v.name == identity {
+			userId = v.userId
+			return
+		}
+	}
+	err = ErrIdentityAuthNotFound
+	return
+}
+
+func (x *dummyUserStore) Close() {
+}
+
+func getUser(users map[UserId]*dummyUser, identity string) (user *dummyUser) {
+	for _ , v := range users {
+		if CanonicalizeIdentity(v.name) == CanonicalizeIdentity(identity) {
+			user = v
+			return
+		}
+	}
+	return
+}
+
+func generateUserId(users map[UserId]*dummyUser) (UserId) {
```

Comment:
Return values should not be in parentheses if there is only one return value. I suspect gofmt would have fixed this up.
But this is a bad design - you are going to be reusing userids. Rather use an incrementing integer so that you never reuse old IDs.